src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -79,6 +83,44 @@ public Task AddComponentAsync(Type componentType, string domElementSelector)
             return RenderRootComponentAsync(componentId);
         }
 
+        protected override void ProcessRenderQueue()
+        {
+            if (_unacknowledgedRenderBatches.Count >= _options.MaxBufferedUnacknowledgedRenderBatches)
+            {
+                // If we got here it means we are at max capacity, so we don't want to actually process the queue,
+                // as we have a client that is not acknowledging render batches fast enough (something we consider needs
+                // to be fast).
+                // The result is somethign as follows:
+                // Lets imagine an extreme case where the server produces a new batch every milisecond.
+                // Lets say the client is able to ACK a batch every 100 miliseconds.
+                // When the app starts the client might see the sequence 0.000->0.{MAXUnacknowledgeRenderBatches} and then
+                // after 100 miliseconds it sees it jump to 0.1xx, then to 0.2xx where xx is something between {0..99} the
+                // reason for this is that the server slows down rendering new batches to as fast as the client can consume
+                // them.
```

Comment:
I don't understand why there would be gaps in the batch number sequence. Don't we only increment the batch number when we call `UpdateDisplay`, which is only when `ProcessRenderQueue` has produced some output?

I would have instead expected:

 * Server renders batches 0..(MaxUnacknowledgedRenderBatches-1)
 * Client receives batch 0..(MaxUnacknowledgedRenderBatches-1)
 * ... wait 100ms
 * Server receives ACKs for all those batches and renders batches MaxUnacknowledgedRenderBatches..(2*MaxUnacknowledgedRenderBatches-1)
 * ...repeat

The fact that components were trying to render more times between steps 1 and 4 wouldn't influence the batch number sequence, would it?

Reply:
No, we stop the render process from updating the display, but not from updating its own state. For example, when you have a timer that ticks every second, and you are in a "paused rendering state", the timer is going to increase the number and try to queue a render each second. 

We can't/won't stop the timer from increasing its count (in fact we physically can't) but we will not queue a new render after the first one (ComponentBase makes sure of that). When the render gets processed it will render with the current timer state, the reason I say a number between 0..99 is because in the best case scenario the timer has increased at 1 tick per second, but there are thead scheduling factors in place in this scenario.

What I'm trying to say is that there's some delay between the render being queued and when the render happens. I'm not sure if I've expressed myself well here. I'll look at the comment in more detail later and see if I can rephrase it in clearer terms.

Reply:
OK, I misunderstood the entire comment block. I thought the numbers were the batch IDs, but now it seems they are the ticking "count" values. So yes I do understand why those would have gaps :)